compiler/src/iree/compiler/Dialect/HAL/Target/CUDA/CUDATarget.cpp

```diff
@@ -86,53 +98,57 @@ static std::string translateModuleToISA(llvm::Module &module,
+    // TODO(thomasraoux): inject __nvvm_reflect-style functions/globals for
+    // bitcode specialization based on the targetMachine and configuration.
+    // These could use any information we have on the IREE side as well as the
+    // TargetMachine instead of just what __nvvm_reflect supports (arch/etc).
+  };
+
+  // Link user modules and libdevice (if required).
+  // Note that linking order matters:
   llvm::Linker linker(module);
+  unsigned linkerFlags =
+      llvm::Linker::LinkOnlyNeeded | llvm::Linker::OverrideFromSrc;
+  if (failed(linkBitcodeObjects(loc, linker, linkerFlags, targetMachine,
+                                objectAttrs, module.getContext(),
+                                specializationCallback))) {
+    return mlir::emitError(loc)
+           << "failed linking in user objects for target triple '"
+           << targetTriple.str() << "'";
+  }
 
-  llvm::MemoryBufferRef bitcodeBufferRef(
-      llvm::StringRef(libdevice_embedded_create()->data,
-                      libdevice_embedded_create()->size),
-      "libdevice bitcode");
-  auto bitcodeModuleValue =
-      llvm::parseBitcodeFile(bitcodeBufferRef, module.getContext());
-  if (!bitcodeModuleValue) {
-    llvm::errs() << "failed to parse CUDA libdevice bitcode: "
-                 << bitcodeModuleValue.takeError();
-    return;
+  if (anyRequiredSymbols(module, "__nv_")) {
+    llvm::MemoryBufferRef bitcodeBufferRef(
+        llvm::StringRef(libdevice_embedded_create()->data,
+                        libdevice_embedded_create()->size),
+        "libdevice.xx.bc");
+    if (failed(linkBitcodeModule(
+            loc, linker, linkerFlags, targetMachine, "libdevice.xx.bc",
+            llvm::parseBitcodeFile(bitcodeBufferRef, module.getContext())))) {
+      return mlir::emitError(loc) << "failed linking in embedded libdevice "
+                                     "bitcode for target triple '"
+                                  << targetTriple.str() << "'";
+    }
   }
-  std::unique_ptr<llvm::Module> bitcodeModule =
-      std::move(bitcodeModuleValue.get());
-  // Ignore the data layout of the module we're importing. This avoids a
-  // warning from the linker.
-  bitcodeModule->setDataLayout(module.getDataLayout());
-  linker.linkInModule(
-      std::move(bitcodeModule), llvm::Linker::Flags::LinkOnlyNeeded,
-      [](llvm::Module &M, const llvm::StringSet<> &GVS) {
-        llvm::internalizeModule(M, [&GVS](const llvm::GlobalValue &GV) {
-          return !GV.hasName() || (GVS.count(GV.getName()) == 0);
-        });
-      });
-}
 
-/// Resolve __nv function by linking libdevice module and run llvm optimization
-/// passes that will inline linked functions and optimize the module.
-static void linkAndOptimize(llvm::Module &module,
-                            llvm::TargetMachine &targetMachine) {
-  if (requiresDeviceLib(module)) linkModule(module);
+  return success();
+}
 
+/// Performs LTO-style optimizations on |module|.
```

Comment:
nit: this comment may be a little confusing as this does much more than just LTO kind of optimizations, we rely on optimizer for quite a lot of other stuff.